plugins/thunderbolt/fu-plugin-thunderbolt.c

```diff
@@ -0,0 +1,736 @@
+						  nwritten,
+						  fw_size - nwritten);
+
+		n = g_output_stream_write_bytes (os,
+						 fw_data,
+						 NULL,
+						 error);
+		if (n < 0) {
+			return FALSE;
+		}
+
+		nwritten += n;
+
+	} while (nwritten < fw_size);
+
+	if (nwritten != fw_size) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_WRITE,
+				     "Could not write all data to nvmem");
+		return FALSE;
+	}
+
+	return g_output_stream_close (os, NULL, error);
+}
+
+typedef struct UpdateData {
+
+	gboolean   have_device;
+	GMainLoop *mainloop;
+	const gchar *target_uuid;
+	guint      timeout_id;
+
+	GHashTable *changes;
+} UpdateData;
+
+static gboolean
+on_wait_for_device_timeout (gpointer user_data)
+{
+	GMainLoop *mainloop = (GMainLoop *) user_data;
+	g_main_loop_quit (mainloop);
+	return FALSE;
+}
+
+static void
+on_wait_for_device_added (FuPlugin    *plugin,
+			  GUdevDevice *device,
+			  UpdateData  *up_data)
+{
+	FuDevice  *dev;
+	const gchar *uuid;
+	const gchar *path;
+	const gchar *version;
+	g_autofree gchar *id = NULL;
+
+	uuid = g_udev_device_get_sysfs_attr (device, "unique_id");
+	if (uuid == NULL)
+		return;
+
+	dev = g_hash_table_lookup (up_data->changes, uuid);
+	if (dev == NULL) {
+		/* a previously unknown device, add it via
+		 * the normal way */
+		fu_plugin_thunderbolt_add (plugin, device);
+		return;
+	}
+
+	/* maybe the device path has changed, lets make sure
+	 * it is correct */
+	path = g_udev_device_get_sysfs_path (device);
+	fu_device_set_metadata (dev, "sysfs-path", path);
+
+	/* make sure the version is correct, might have changed
+	 * after update. */
+	version = g_udev_device_get_sysfs_attr (device, "nvm_version");
+	fu_device_set_version (dev, version);
+
+	id = fu_plugin_thunderbolt_gen_id (device);
+	fu_plugin_cache_add (plugin, id, dev);
+
+	g_hash_table_remove (up_data->changes, uuid);
+
+	/* check if this device is the target*/
+	if (g_str_equal (uuid, up_data->target_uuid)) {
+		up_data->have_device = TRUE;
+		g_debug ("target (%s) re-appeared", uuid);
+		g_main_loop_quit (up_data->mainloop);
+	}
+}
+
+static void
+on_wait_for_device_removed (FuPlugin    *plugin,
+			    GUdevDevice *device,
+			    UpdateData *up_data)
+{
+	g_autofree gchar *id = NULL;
+	FuDevice  *dev;
+	const gchar *uuid;
+
+	id = fu_plugin_thunderbolt_gen_id (device);
+	dev = fu_plugin_cache_lookup (plugin, id);
+
+	if (dev == NULL)
+		return;
+
+	fu_plugin_cache_remove (plugin, id);
+	uuid = fu_device_get_id (dev);
+	g_hash_table_insert (up_data->changes,
+			     (gpointer) uuid,
+			     g_object_ref (dev));
+
+	/* check if this device is the target*/
+	if (g_str_equal (uuid, up_data->target_uuid)) {
+		up_data->have_device = FALSE;
+		g_debug ("target (%s) disappeared",
+			 uuid);
+	}
+}
+
+static void
+on_wait_for_device_notify (FuPlugin    *plugin,
+			   GUdevDevice *device,
+			   const char  *action,
+			   gpointer    user_data)
+{
+	UpdateData *up_data = (UpdateData *) user_data;
+
+	if (g_str_equal (action, "add")) {
```

Comment:
Nope, it is being called from `udev_event_cb`, which checks for that. I will add a comment.